lmnet/lmnet/post_processor.py

```diff
@@ -419,3 +419,74 @@ class Softmax(Processor):
     def __call__(self, outputs, **kwargs):
         results = _softmax(outputs)
         return dict({'outputs': results}, **kwargs)
+
+
+class FormatJoints(Processor):
+
+    def __init__(self, num_dimensions=2, stride=2, confidence_threshold=0.1):
+        """
+        Args:
+            num_dimensions: int, it only supports 2 for now.
+            stride: int, stride = image_height / heatmap_height.
+            confidence_threshold: float, [0, 1], ratio of threshold_value and max_value.
+        """
+        self.num_dimensions = num_dimensions
+        self.stride = stride
+        self.confidence_threshold = confidence_threshold
+
+    def __call__(self, outputs, *args, **kwargs):
+        """
+        Args:
+            outputs: output heatmaps, a numpy array of shape (batch_size, height, width, num_joints).
+
+        Returns:
+            all args (dict):
+                outputs: joints, a numpy array of shape (batch_size, num_joints, num_dimensions + 1).
+
+        """
+
+        batch_size = outputs.shape[0]
+        num_joints = outputs.shape[3]
+
+        joints = np.zeros((batch_size, num_joints, self.num_dimensions + 1), dtype=np.float32)
+
+        for i in range(batch_size):
+            joints[i] = gaussian_heatmap_to_joints(outputs[i],
+                                                   num_dimensions=self.num_dimensions,
+                                                   stride=self.stride,
+                                                   confidence_threshold=self.confidence_threshold)
+
+        return dict({'outputs': joints}, **kwargs)
+
+
+def gaussian_heatmap_to_joints(heatmap, num_dimensions=2, stride=2, confidence_threshold=0.1):
+    """
+    Extract joints from gaussian heatmap. Current version only supports 2D pose estimation.
+    Args:
+        heatmap: a numpy array of shape (height, width, num_joints).
+        num_dimensions: int, it only supports 2 for now.
+        stride: int, stride = image_height / heatmap_height.
+        confidence_threshold: float, [0, 1], ratio of threshold_value and max_value.
+
+    Returns:
+        joints: a numpy array of shape (num_joints, num_dimensions + 1).
+
+    """
+
+    height, width, num_joints = heatmap.shape
+
+    threshold_value = 10 * confidence_threshold
```

Comment:
This `10` looks a little sudden. How about using 10 times larger confidence threshold?

Reply:
It is sudden... I got `10` just from tuning... I set it to 10 by default in pre_processor.
https://github.com/blue-oil/blueoil/blob/4928b8ed05023b3fafe1f874c93d2e587c637a6b/lmnet/lmnet/pre_processor.py#L451
Two solutions here, I can hard-code both or add a `max_value` argument here as pre_processor.

Which one do you think is better? 🤔

Reply:
Thanks, I understood better than before. I recommend to omit this `10` (and `max_value` also.) A user need to optmize threshold value (if needed) anyhow, hence adding a hyperparameter that only affects the threshold value linearly only increases the effort for users.

Another option is that you can extract the max_value automatically from a heat map, but personally I don't prefer to add such a feature which might not be trivial from user's perspective.

Reply:
Thanks. Fixed.

Reply:
There seems a miscommunication... 😢 I meant all `max_value` (except the line 492&493) should be rtemoved. I'll explain why I'm thinking so.

- From the docstring, it would be hard to understand what will happen if a user change the `confindence_threshold`, since none of `threshold_value` and `max_value` is explained.
- `max_value` in the line 484 doesn't help a user. The result is exactly the same as multiplying the `confidence_threshold` by 10. User will be confused because they cannot know that actual threshold value is multiplied by 10, until they read the code.

Please keep in mind, we should write code that the user can guess the behavior by looking at the API. I'm thinking that this point is very important for your growth as a software engineer. If you feel that your understanding is not enough, please ask me a question offline at any time 😉

Reply:
It was confusing that I used `max_value` twice.. I have fixed it. Thanks! 😄

Reply:
It's not a problem of using `max_value` twice. Could you explain why do you want multiply `10` to heatmap?

Reply:
Since I found that quantized networks are hard to regress a heatmap directly ranged from 0 to 1. So I multiply it by `10` and it works. `10` is a tuned result.

Reply:
Then, `max_value` would not be an appropriate variable name. How about `heatmap_scaling_coefficient` or `heatmap_scaling_factor`?

Reply:
Yes! it is. I'm sorry for this bad name. I was confused to choose this name! `heatmap_scaling_factor` should be better I think. Do I need to add a redundant variable or just change the comments.